Make building much easier and flexible

TheRealKeto opened the pull request:
This PR makes building the project much easier and flexible by using a Makefile. This way, a two-liner can be used in order to do all the work (rather than a long command with arguments).

The workflow was also changed in order to use the Makefile, and the new commands were reflected on the README. Specific changes have been documented in the commit message.

TheRealKeto committed:
Make building easier with a Makefile

TheRealKeto committed:
Force workflow to use added Makefile

- Force CI workflow to use Makefile
- Use new method for copying resource files

TheRealKeto committed:
Further changes and fixes

- Don't use CC variable; you can't overwrite this on Linux
- Fix a small logical error when cloning emcc
- Fix logical error in EMCC_FLAGS flags variable

TheRealKeto committed:
Change building instructions in README

TheRealKeto commented:
I'm marking this as a draft since I realized that this can be done much better; once I implement those changes (and there are no PRs for the workflow), I will change the status of the PR.

TheRealKeto committed:
Further changes and fixes

- Actually preform a check for Windows (and set OS specific vars)
- Rewrite most strucutre of Makefile; one function handles all builds
- Copy resources to correct folder with compiled binary
- Handle copying web build files within the Makefile instead

TheRealKeto committed:
Adapt to changes from previous commit

- Allow Makefile to handle copying web build files
- Update APT sources before installing packages
- Enable emsdk env when compiling, not before
- Change name and directories for artifact uploads
- Shorten command for Linux dependencies

TheRealKeto committed:
Update compiling notes

TheRealKeto commented:
This is ready, just need to update compiling notes again; Linux and web builds compile successfully.

mandar1jn commented on .github/workflows/build.yaml:
```diff
@@ -1,43 +1,58 @@
+name: CI
 
-on: [push, pull_request]
+on:
+  # Trigger on all pushes and PRs
+  push:
+  pull_request:
+
+  # Allow manually running workflow
+  workflow_dispatch:
```
```suggestion
on: [push, pull_request, workflow_dispatch]
```
As the writer of the original workflow I would like to request this change. It's essentially the same, but only uses one line of code instead of 7.

TheRealKeto replied:
A specific PR (which disables the workflow for small commits, such as a README change) uses my current format, so it's not logical to simply use a list after that PR is merged. This PR will obviously reflect those changes once merged.

mandar1jn replied:
Damn. I totally forgot. I wrote that PR. I didn't keep merge issues in mind while making this suggestion

mandar1jn reviewed (approved):
This should be merged. This makefile is very handy and as the author of the workflow I can say that this is way beter

TheRealKeto committed:
Further workflow improvements

- Add individual names to jobs
- Cache emscripten setup (rather than always setting it up)

TheRealKeto committed:
QOL update to Makefile

- Adapt to using CXX instead of specifying g++
- Change variables names to be more fitting
- Create prep rules for both Linux and web builds

TheRealKeto committed:
Add clean Makefile rule

- Rename some Makefile rules
- Add clean Makefile rule
- Add RES variable, use more wildcards

TheRealKeto commented:
This is ready now. Everything is as fast as it will ever be. 

I managed to make the workflow cache emscripten so that it doesn't need to be setup every time (for the web-build). This increased the speed of the workflow after its first run.
